Add to Featured Product block option to remove custom image

tomasztunik opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This adds ability to remove custom image from Featured Product block. Removing custom image will reset it back to the default product image if available. This follows solution from #5719 which added this option to Featured Category block.

<!-- Reference any related issues or PRs here -->
Fixes #5725

<!-- Don't forget to update the title with something descriptive. -->
<!-- If your pull request implements a feature flag, make sure you update [this doc](../docs/blocks/features-and-blocks-behind-a-flag.md) -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)


**UPDATE**: "Remove" label replaced with "Reset"
### Screenshots
![Screenshot 2022-02-18 at 00 02 37](https://user-images.githubusercontent.com/112270/154586939-a9802d26-a5fe-45a4-9f7a-34a74b0930b5.jpg)

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

### Manual Testing

How to test the changes in this Pull Request:

1. Add a Featured Product block to a page, post, or template.
2. Add custom background media.
3. The button to remove the image should be available.
4. If clicked it should remove the custom media.
5. If the product had an image it should display it again.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [x] Same as above, or
* [ ] See steps below.
### Note

misaligned "Shop Now" is an unrelated issue

### Changelog

> Featured Product block: add the ability to reset to a previously set custom background image.

tomasztunik committed:
adds toolbar option to remove custom image

removing custom image will reset it back to the default
product image if available

dinhtungdu reviewed (commented):
@tomasztunik This is working great on my end. I have two minor issues, otherwise, this is looking good to me.

- For the button label, I think `Reset` is better to describe the function of that button than `Remove`. When users click that button, if the product has a featured image, the button doesn't really "remove" the image from the users' perspective. If the label of the button is `Remove custom featured image`, then using `Remove` makes sense, but it'd be too long.
- For the buttons group containing only icons, we don't need a separator between buttons, but for buttons with text labels, I think a separator should be used to distinguish buttons.

![image](https://user-images.githubusercontent.com/5423135/154668518-86f1cd4e-eddd-41cb-b953-82dcc7a85ea5.png)

tomasztunik commented:
Valid points @dinhtungdu. Tried to just implement existing flow but maybe we could improve this and Featured Category blocks.

With the separator I think it wasn't added there to highlight that these two actions are related to the same background image selector context. Not sure what is the definitive way to go here.

In general I feel like the dominating "Replace" button especially if category on Featured Category block or Product is selected these seem to refer to the context of a block so feature or product rather than the image. 

Maybe we should use icons there? Went with undo as trash felt too destructive and "remove" like a stop sign :) 

![Screenshot 2022-02-18 at 12 35 09](https://user-images.githubusercontent.com/112270/154675124-93a3f22b-1e55-4260-9cf9-86500909873a.jpg)

![Screenshot 2022-02-18 at 12 33 51](https://user-images.githubusercontent.com/112270/154675148-16b787e6-d227-4675-9441-d539b4c0f1c8.jpg)

VS current with text labels

![Screenshot 2022-02-18 at 12 38 43](https://user-images.githubusercontent.com/112270/154675926-d2ffb99e-13e8-41cb-ba2e-2f7fefeec631.jpg)

@vivialice what do you think? Should we stick to current solution? go with icons? else?

(Used Featured Category block as ref as I had it at hand, but they follow the same patterns)

vivialice commented:
Hey team, I like these solutions and I think it makes sense to make that portion of the toolbar more useful. I like the icon solution but I think it could be a safer bet to follow what Gutenberg have in image blocks and keep `replace` rather than the picture icon for consistency. I like the `reset` arrow icon but since we are adding something "new" we might need to check with Gutenberg if that is okay. I agree we should group these two functions in the toolbar. 

Another idea:
<img width="280" alt="Screenshot 2022-02-18 at 12 34 15" src="https://user-images.githubusercontent.com/9888201/154683757-492e7bd3-53ea-4238-985e-31781f1c50d2.png">
This is from the Cover block. There is a `clear media` CTA in the sidebar settings but we could change the label to `reset changes`.

tomasztunik committed:
update copy as per discussion

went with just Reset as it is commonly used in WordPress
and will have translation available

tomasztunik commented:
Updated with "Reset" label and retained no buttons separation to keep the context of working with background media. I've seen this mentioned elsewhere that it would be nice if these were more unified with Cover block and I think any bigger changes for Categories and Product blocks should probably include this consideration. The UI on editing cover block is more clear and more user friendly and provides more options in the sidebar than we have available here.

dinhtungdu reviewed (approved):
LGTM :shipit:

Aljullu commented:
@tomasztunik could you create a good-first-issue (or directly a PR) to make sure the button label is aligned between both blocks? Currently it's [_Reset_ in Featured Product](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/82ff784f25a78b1d3557a73c23eeee40dfb417cb/assets/js/blocks/featured-product/block.js#L161) but [_Remove_ in Featured Category](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/82ff784f25a78b1d3557a73c23eeee40dfb417cb/assets/js/blocks/featured-category/block.js#L113), ideally I think they should be the same. Do you agree?

tomasztunik commented:
@Aljullu was planning to but the alignment issue got fixed by https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5867 already!

Regarding the reset/remove - here it resets to product image - so we still have an image, there it removes the image completely and uses a background color as selected with styles. Should we unify these despite slightly different behaviour?

Aljullu commented:
> there it removes the image completely and uses a background color as selected with styles

Does it? It seems to reset the image to whatever the category image is for me:

https://user-images.githubusercontent.com/3616980/155307005-ac5f6cd9-08eb-4b5e-9b8b-b6ad7c5d02d4.mp4

(Notice that if the category doesn't have an image, then yes, a background color is used)

tomasztunik commented:
Ah I see, had a category without image when testing. Will update it to reflect that if image is present it will say reset else remove? or always reset?

Aljullu commented:
> Ah I see, had a category without image when testing. Will update it to reflect that if image is present it will say reset else remove? or always reset?

IMO always reset is good enough. To me, it makes sense in both contexts.
